### PR TITLE
PLATFORM-1267: add etag to response headers

### PR DESCRIPTION
**Pull request by pchojnacki**
This PR only adds etag header and not If-None-Match check that would return 304 if etags match.

My Vignette/Clojure fu needs some more fine tuning for that.
In next PR I'd also like to HEAD method so that it will return the same set of headers as GET.

/cc: @nmonterroso @drsnyder

### CODE REVIEW

**Comment by nmonterroso** [comment]
LGTM :ship:

**Comment by ljagiello** [comment]
Request:
```
➜  ~  curl -svo /dev/null http://vignette.wikia-dev.com/muppet/images/3/35/Slider-20130415-movienews.jpg/revision/latest/scale-to-width-down/670\?cb\=20130416002146 -v -H "If-None-Match: c139ec491ac28f105dcd598e95e50292"
[…]
> GET /muppet/images/3/35/Slider-20130415-movienews.jpg/revision/latest/scale-to-width-down/670?cb=20130416002146 HTTP/1.1
> User-Agent: curl/7.37.1
> Host: vignette.wikia-dev.com
> Accept: */*
> If-None-Match: c139ec491ac28f105dcd598e95e50292
>
< HTTP/1.1 304 Not Modified
< Content-Disposition: inline; filename="Slider-20130415-movienews.jpg"
< X-Thumbnailer: Vignette
< Content-Type: image/jpeg
< Cache-Control: public, max-age=31536000
< X-Surrogate-Key: 715f3ad1e8440982e1e1d4b6c8fc11806a168aec
< ETag: c139ec491ac28f105dcd598e95e50292
* Server Jetty(9.2.z-SNAPSHOT) is not blacklisted
< Server: Jetty(9.2.z-SNAPSHOT)
< Accept-Ranges: bytes
< Accept-Ranges: bytes
< X-Cacheable: NO:Not-Cacheable
< Accept-Ranges: bytes
< Date: Fri, 10 Jul 2015 18:19:47 GMT
< Connection: keep-alive
< X-Served-By: dev-vignette, cache-wk-sjc3160-WIKIA, cache-lax1428-LAX
< X-Cache: ORIGIN, MISS, MISS
< X-Cache-Hits: ORIGIN, 0, 0
< X-Timer: S1436552387.671800,VS0,VE116
< Vary: Accept-Encoding
```

From `dev-vignette` perspective:
```
T 10.8.66.62:55274 -> 10.8.64.105:5000 [AP]
GET /muppet/images/3/35/Slider-20130415-movienews.jpg/revision/latest/scale-to-width-down/670?cb=20130416002146 HTTP/1.1.
User-Agent: curl/7.37.1.
Host: vignette.wikia-dev.com.
Accept: */*.
If-None-Match: c139ec491ac28f105dcd598e95e50292.
Fastly-Client-IP: 24.130.195.44.
X-1955fb8700b3b47cf7f3a98806ee1881: 24.130.195.44.
X-Forwarded-For: 24.130.195.44.
X-Varnish: 2718870193.
X-Timer: S1436552208.598448,VS0,VS0.
X-Slot: 1.
X-Slot-Host: slot1.wikia.com.
X-Set-Beacon-Cookie: 0.
X-Varnish: 2514762071.
Fastly-FF: cache-sjc3120-SJC, cache-wk-sjc3161-WIKIA.
X-Varnish: 2292310147.
.


T 127.0.0.1:55362 -> 127.0.0.1:6081 [AP]
GET http://muppet.dfs/images/3/35/Slider-20130415-movienews.jpg HTTP/1.1.
Host: muppet.dfs.
Authorization: AWS C82HQF6UMB68ZTAYIN1V:d1QCrzvIIooZmXUEN4mI7iXrjDg=.
User-Agent: aws-sdk-java/1.7.5 Linux/3.13.0-24-generic Java_HotSpot(TM)_64-Bit_Server_VM/25.31-b07/1.8.0_31.
Date: Fri, 10 Jul 2015 18:16:48 GMT.
Content-Type: application/x-www-form-urlencoded; charset=utf-8.
Proxy-Connection: Keep-Alive.
.


T 10.8.64.105:55492 -> 10.8.68.43:9060 [AP]
GET /muppet/images/3/35/Slider-20130415-movienews.jpg HTTP/1.1.
Host: muppet.dfs.
Authorization: AWS C82HQF6UMB68ZTAYIN1V:d1QCrzvIIooZmXUEN4mI7iXrjDg=.
User-Agent: aws-sdk-java/1.7.5 Linux/3.13.0-24-generic Java_HotSpot(TM)_64-Bit_Server_VM/25.31-b07/1.8.0_31.
Date: Fri, 10 Jul 2015 18:16:48 GMT.
Content-Type: application/x-www-form-urlencoded; charset=utf-8.
Proxy-Connection: Keep-Alive.
X-Varnish: 154926638.
.


T 10.8.64.105:5000 -> 10.8.66.62:55274 [A]
HTTP/1.1 200 OK.
X-Cache-Hits: ORIGIN.
Content-Disposition: inline; filename="Slider-20130415-movienews.jpg".
X-Thumbnailer: Vignette.
Varnish-Logs: vignette.
Surrogate-Key: 715f3ad1e8440982e1e1d4b6c8fc11806a168aec.
X-Cache: ORIGIN.
Content-Type: image/jpeg.
Cache-Control: public, max-age=31536000.
X-Surrogate-Key: 715f3ad1e8440982e1e1d4b6c8fc11806a168aec.
ETag: c139ec491ac28f105dcd598e95e50292.
X-Served-By: dev-vignette.
Content-Length: 73658.
Connection: close.
Server: Jetty(9.2.z-SNAPSHOT).
```

Vignette is responding 304 but under the hood is doing a full GET request to dfs. I would expect to see HEAD request to dfs and based on etag answer 304 or in case of mismatch GET for a file. In current configuration we're generating tons of internal traffic for super light requests.

**Comment by ljagiello** [comment]
@pchojnacki I'm not sure if next PR with HEAD method will cover this case but:

After such request:
```
➜  ~  curl -I -svo /dev/null http://dfs-s1/poznan/pl/images/0/00/IMG_2721.jpg
> HEAD /poznan/pl/images/0/00/IMG_2721.jpg HTTP/1.1
> User-Agent: curl/7.37.1
> Host: dfs-s1
> Accept: */*
>
< HTTP/1.1 200 OK
* Server nginx is not blacklisted
< Server: nginx
< Date: Fri, 10 Jul 2015 18:13:59 GMT
< Content-Type: image/jpeg
< Content-Length: 169222
< Connection: keep-alive
< Accept-Ranges: bytes
< Last-Modified: Fri, 07 Feb 2014 13:40:03 GMT
< ETag: "7e3078bc3ba9352f3e05de4b338611ad"
< x-amz-meta-sha1base36: hjgvqp3ha6vpaz08p9tav3esx1gsfyv
< Expires: Sat, 09 Jul 2016 18:13:59 GMT
< Cache-Control: max-age=31536000
< Pragma: public
< Cache-Control: public
< X-Served-By: dfs-s1
```

From `dev-vignette` perspective traffic looks like that:
```
T 10.8.66.62:55274 -> 10.8.64.105:5000 [AP]
GET /muppet/images/3/35/Slider-20130415-movienews.jpg/revision/latest/scale-to-width-down/670?cb=20130416002146 HTTP/1.1.
User-Agent: curl/7.37.1.
Host: vignette.wikia-dev.com.
Accept: */*.
If-None-Match: c139ec491ac28f105dcd598e95e50292.
Fastly-Client-IP: 24.130.195.44.
X-1955fb8700b3b47cf7f3a98806ee1881: 24.130.195.44.
X-Forwarded-For: 24.130.195.44.
X-Varnish: 2718870193.
X-Timer: S1436552208.598448,VS0,VS0.
X-Slot: 1.
X-Slot-Host: slot1.wikia.com.
X-Set-Beacon-Cookie: 0.
X-Varnish: 2514762071.
Fastly-FF: cache-sjc3120-SJC, cache-wk-sjc3161-WIKIA.
X-Varnish: 2292310147.
.


T 127.0.0.1:55362 -> 127.0.0.1:6081 [AP]
GET http://muppet.dfs/images/3/35/Slider-20130415-movienews.jpg HTTP/1.1.
Host: muppet.dfs.
Authorization: AWS C82HQF6UMB68ZTAYIN1V:d1QCrzvIIooZmXUEN4mI7iXrjDg=.
User-Agent: aws-sdk-java/1.7.5 Linux/3.13.0-24-generic Java_HotSpot(TM)_64-Bit_Server_VM/25.31-b07/1.8.0_31.
Date: Fri, 10 Jul 2015 18:16:48 GMT.
Content-Type: application/x-www-form-urlencoded; charset=utf-8.
Proxy-Connection: Keep-Alive.
.


T 10.8.64.105:55492 -> 10.8.68.43:9060 [AP]
GET /muppet/images/3/35/Slider-20130415-movienews.jpg HTTP/1.1.
Host: muppet.dfs.
Authorization: AWS C82HQF6UMB68ZTAYIN1V:d1QCrzvIIooZmXUEN4mI7iXrjDg=.
User-Agent: aws-sdk-java/1.7.5 Linux/3.13.0-24-generic Java_HotSpot(TM)_64-Bit_Server_VM/25.31-b07/1.8.0_31.
Date: Fri, 10 Jul 2015 18:16:48 GMT.
Content-Type: application/x-www-form-urlencoded; charset=utf-8.
Proxy-Connection: Keep-Alive.
X-Varnish: 154926638.
.


T 10.8.64.105:5000 -> 10.8.66.62:55274 [A]
HTTP/1.1 200 OK.
X-Cache-Hits: ORIGIN.
Content-Disposition: inline; filename="Slider-20130415-movienews.jpg".
X-Thumbnailer: Vignette.
Varnish-Logs: vignette.
Surrogate-Key: 715f3ad1e8440982e1e1d4b6c8fc11806a168aec.
X-Cache: ORIGIN.
Content-Type: image/jpeg.
Cache-Control: public, max-age=31536000.
X-Surrogate-Key: 715f3ad1e8440982e1e1d4b6c8fc11806a168aec.
ETag: c139ec491ac28f105dcd598e95e50292.
X-Served-By: dev-vignette.
Content-Length: 73658.
Connection: close.
Server: Jetty(9.2.z-SNAPSHOT).
```

So we're doing full GET when simple HEAD would be enough to compare etags and after mismatch there should be a full GET.

**Comment by pchojnacki** [comment]
I'm not sure I'm following you.
But as far I understand how Varnish handles ETag. It will never by itself set IfModifiedSince header.
It just compares "StoredOrRetrieved" document ETag to decide what to return to user. 
```
https://github.com/varnish/Varnish-Cache/blob/c037d5b21a345a3eead0be9daaee221523a20a81/bin/varnishd/cache/cache_rfc2616.c#L236
https://github.com/varnish/Varnish-Cache/search?utf8=%E2%9C%93&q=H_If_None_Match&type=Code
```
Vingette OTOH requires some rework with how it handles HEAD requests (head and 304 response look almost the same and HEAD right now returns limited set of headers ) so it wouldn't retrieve whole object from store. Also IDK if Vignette s3 libs support IfNoneMatch semantics and if not two requests would need to be made to implement IfNoneMatch optimal behavior.

I think just passing ETag allong for Varnish/Fastly to use is enough to bring in 99% of possible cost optimization.

**Comment by pchojnacki** [comment]
@ljagiello btw you do know this is Public repo so anyone can see your comments ?

**Comment by ljagiello** [comment]
Ok stripped sensitive informations.

It's not connected with Varnish/Fastly at all. It's a problem how Vignette generate 304 answer. In first request I'm asking is file still the same `"If-None-Match: c139ec491ac28f105dcd598e95e50292"` in return Vignette is answering with 304 and of course etag is `c139ec491ac28f105dcd598e95e50292` and that part if perfectly fine.

The problem is how vignette is generating that answer. At second part you can see it's doing normal GET request, download entire file, generate result, calculate etag.… and that part IMHO is wrong. It should be done with w HEAD request and based on result we should return 304 or actually do a GET (thumb is not there anymore). With a current logic for every 304 request we will download and generate thumb.

File: src/vignette/storage/local.clj [hunk 1]
```diff
@@ -3,6 +3,7 @@
             [clojure.java.shell :as sh]
             [pantomime.mime :refer [mime-type-of]]
             [vignette.media-types :as mt]
+            [digest]
```
Comment (drsnyder) [inline src/vignette/storage/local.clj]: Minor: We normally either `:refer :all` or `:refer [fn-i-want-to-use other-fn-i-want-to-use]` to make it clear what we are using.

**Pull request merged**